sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,545 @@
+    expected within the dataset. Lower eps values can be used after
+    OPTICS is run the first time, with fast returns of labels. Default
+    value of "np.inf" will identify clusters across all scales; reducing
+    eps will result in shorter run times.
+    min_samples : int, optional
+    The number of samples in a neighborhood for a point to be considered
+    as a core point.
+    metric : string or callable, optional
+    The distance metric to use for neighborhood lookups. Default is
+    "minkowski". Other options include “euclidean”, “manhattan”,
+    “chebyshev”, “haversine”, “seuclidean”, “hamming”, “canberra”,
+    and “braycurtis”. The “wminkowski” and “mahalanobis” metrics are
+    also valid with an additional argument.
+
+    Attributes
+    ----------
+    `core_sample_indices_` : array, shape = [n_core_samples]
+        Indices of core samples.
+    `labels_` : array, shape = [n_samples]
+        Cluster labels for each point in the dataset given to fit().
+        Noisy samples are given the label -1.
+    `reachability_` : array, shape = [n_samples]
+        Reachability distances per sample
+    `ordering_` : array, shape = [n_samples]
+        The cluster ordered list of sample indices
+    `core_dists_` : array, shape = [n_samples]
+        Distance at which each sample becomes a core point.
+        Points which will never be core have a distance of inf.
+
+    References
+    ----------
+    Ankerst, Mihael, Markus M. Breunig, Hans-Peter Kriegel, and Jörg Sander.
+    "OPTICS: ordering points to identify the clustering structure." ACM SIGMOD
+    Record 28, no. 2 (1999): 49-60.
+    """
+
+    def __init__(self, eps=np.inf, min_samples=5,
+                 metric='euclidean', **kwargs):
+        self.eps = eps
+        self.min_samples = min_samples
+        self.metric = metric
+        self.processed = False
+
+    def fit(self, X, y=None):
+        """Perform OPTICS clustering
+
+        Extracts an ordered list of points and reachability distances, and
+        performs initial clustering using 'eps' distance specified at OPTICS
+        object instantiation.
+
+        Parameters
+        ----------
+        X : array [n_samples, n_features]"""
+
+        X = check_array(X)
+
+        # Check for valid n_samples relative to min_samples
+        if self.min_samples > len(X):
+            print("Number of training samples must be greater than")
+            print("min_samples used for clustering")
+            return
+
+        self.tree = SetOfObjects(X, self.metric)
+        _prep_optics(self.tree, self.min_samples)
+        _build_optics(self.tree, self.eps * 5.0)
+        self._index = self.tree._index[:]
+        self.reachability_ = self.tree.reachability_[:]
+        self.core_dists_ = self.tree.core_dists_[:]
+        self._cluster_id = self.tree._cluster_id[:]
+        self._is_core = self.tree._is_core[:]
+        self.ordering_ = self.tree.ordering_[:]
+        self.extract_auto()
```

Comment:
This may actually make sense to be public, as it has the documentation on the hyper parameters for the extraction-- these can be passed by kwargs, although I'm not sure if this is a no-no as far as the API is concerned...